py-basemap: add v1.4.1

LaurentChardon opened the pull request:
* py-basemap: remove broken v1.2.1
* py-basemap: add v1.4.1
* py-basemap: add hires variant
* py-basemap: change installation method to pip

LaurentChardon committed:
py-basemap: add v1.4.1

* py-basemap: remove broken v1.2.1
* py-basemap: add v1.4.1
* py-basemap: add hires variant
* py-basemap: change installation method to pip

LaurentChardon committed:
Merge branch 'develop' into py-basemap

tldahlgren commented on var/spack/repos/builtin/packages/py-basemap/package.py:
```diff
@@ -9,33 +9,37 @@ class PyBasemap(PythonPackage):
 
     license("MIT")
 
-    version("1.2.1", sha256="3fb30424f18cd4ffd505e30fd9c810ae81b999bb92f950c76553e1abc081faa7")
```
Existing versions listed in a package are *usually* deprecated. See the docs at https://spack.readthedocs.io/en/latest/packaging_guide.html#deprecating-old-versions for more information.

tldahlgren replied:
You could add the URL as an argument to this directive to retain it, e.g.,:

```
    version(
        "1.2.1",
        sha256="3fb30424f18cd4ffd505e30fd9c810ae81b999bb92f950c76553e1abc081faa7",
        url="https://github.com/matplotlib/basemap/archive/v1.2.1rel.tar.gz",
    )
```

LaurentChardon replied:
I removed it because it was broken, i couldn't compile it under any platform. I didn't see value in spending time making that very old version work because no package depends on it besides mlperf-deepcam, and it works fine with version 1.4.1.

tldahlgren commented on var/spack/repos/builtin/packages/py-basemap/package.py:
```diff
@@ -9,33 +9,37 @@ class PyBasemap(PythonPackage):
     license("MIT")
 
-    version("1.2.1", sha256="3fb30424f18cd4ffd505e30fd9c810ae81b999bb92f950c76553e1abc081faa7")
+    version("1.4.1", sha256="730b1e2ff5eb31c73680bd8ebabc6b11adfc587cfa6832c528a8a82822e5a490")
```
Where did the sha256 come from?

According to https://pypi.org/project/basemap/#basemap-1.4.1.zip, it should be `6822d3d08c11cdc443e3ed01d61d512e7cf62d4b40bbc1d235f0a898f2c54a7a`.

tldahlgren replied:
Okay. I think I see where the sha256 you used is coming from. Please remove `url_for_version()` since it's interfering with use of `pypi.org`.

tldahlgren replied:
Once it's removed, I can confirm the official sha256.

LaurentChardon replied:
OK, I will remove it.  

For information, the sha256sum is computed by spack with this PR. It uses the tar.gz file, not the zip.

```
$ spack checksum -l py-basemap
==> Found 1 version of py-basemap
==> Fetching https://github.com/matplotlib/basemap/archive/v1.2.0rel.tar.gz

    version("1.2.0", sha256="bd5bf305918a2eb675939873b735238f9e3dfe6b5c290e37c41e5b082ff3639a")

$ git switch py-basemap
Switched to branch 'py-basemap'
Your branch is behind 'origin/py-basemap' by 2 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)

$ spack checksum -l py-basemap
==> Found 1 version of py-basemap
==> Fetching https://github.com/matplotlib/basemap/archive/v1.4.1.tar.gz

    version("1.4.1", sha256="730b1e2ff5eb31c73680bd8ebabc6b11adfc587cfa6832c528a8a82822e5a490")
```

tldahlgren reviewed (commented):
Unclear what is going on but it is taking a long time for me to simply stage the package to check  dependencies.

tldahlgren commented on var/spack/repos/builtin/packages/py-basemap/package.py:
```diff
@@ -44,3 +48,14 @@ def url_for_version(self, version):
         env.set("GEOS_DIR", self.spec["geos"].prefix)
+
+    def install(self, spec, prefix):
+        with working_dir("packages/basemap"):
+            pip("install", ".", "--prefix={0}".format(prefix))
+
+        with working_dir("packages/basemap_data"):
+            pip("install", ".", "--prefix={0}".format(prefix))
+
+        if "+hires" in spec:
+            with working_dir("packages/basemap_data_hires"):
+                pip("install", ".", "--prefix={0}".format(prefix))
```
🤔 Not sure about this approach. I'm not aware of seeing any other python packages  using `pip` to install their software.

Based on the `README.md`, it would be better to separate the packages and put a dependency for `py-basemap` on `py-basemap-data`.

```
The package is available in PyPI and can be installed with [`pip`]:

python -m pip install basemap-data
```

tldahlgren replied:
Given the dependencies, I wonder if it would be better to define a separate package for `py-basemap-data`.

tldahlgren replied:
Same appears to be true for `py-basemap-data-hires`:

```
The package is available in PyPI and can be installed with [`pip`]:

python -m pip install basemap-data-hires
```

LaurentChardon replied:
About the approach, my thinking was that using pyproject.toml was more future-proof than setup.py, but I can modify that.

I disagree that it would be better to have three separate packages for basmap, basmap-data and basemap-data-hires, because it's all the exact source code package. To me, the variant approach for whether or not you want hires data is better, simpler and safer in terms of not getting divergence. That bing said, if there is a good reason for separating them, I'll do it, just let me know.

tldahlgren commented on var/spack/repos/builtin/packages/py-basemap/package.py:
```diff
@@ -9,33 +9,37 @@ class PyBasemap(PythonPackage):
+    depends_on("geos", type=("build", "run"))
 
     def url_for_version(self, version):
-        if version >= Version("1.2.0"):
+        if version >= Version("1.3.2"):
+            return "https://github.com/matplotlib/basemap/archive/v{0}.tar.gz".format(version)
+        elif version >= Version("1.2.0"):
             return "https://github.com/matplotlib/basemap/archive/v{0}rel.tar.gz".format(version)
         else:
             return "https://downloads.sourceforge.net/project/matplotlib/matplotlib-toolkits/basemap-{0}/basemap-{0}.tar.gz".format(
```
I'm pretty sure this is no longer needed when switching from `url` to `pypi`.

LaurentChardon replied:
I removed 1.2.1 because it was broken and I preferred spending time making the latest version work rather than a much older version.

I left the older version formats in case someone needed to reintroduce an older version, but looking at it now, I agree that it's silly, they could just look in the git history. I'll remove the older version formats.

About the three separate packages, see my thoughts on that in https://github.com/spack/spack/pull/48548#discussion_r1914582740

tldahlgren commented on var/spack/repos/builtin/packages/py-basemap/package.py:
```diff
@@ -9,33 +9,37 @@ class PyBasemap(PythonPackage):
     # to make mpl_toolkits a namespace package that can span multiple
     # directories (i.e., matplotlib and basemap)
     depends_on("py-setuptools", type=("build", "run"))
     depends_on("py-numpy@1.2.1:", type=("build", "run"))
```
Given the python range constraint on line 24 and the contents of `basemap`'s `requirements.txt`

```
numpy >= 1.11, < 1.12; python_version == "2.6"
numpy >= 1.16, < 1.17; python_version == "2.7"
numpy >= 1.11, < 1.12; python_version == "3.2"
numpy >= 1.11, < 1.12; python_version == "3.3"
numpy >= 1.15, < 1.17; python_version == "3.4"
numpy >= 1.16, < 1.19; python_version == "3.5"
numpy >= 1.19, < 1.20; python_version == "3.6"
numpy >= 1.21, < 1.22; python_version == "3.7"
numpy >= 1.21, < 1.27; python_version >= "3.8"
```

this should be:

```suggestion
    depends_on("py-numpy@1.21:1.26", type=("build", "run"))
```

LaurentChardon replied:
I agree, I'll fix that.

tldahlgren commented on var/spack/repos/builtin/packages/py-basemap/package.py:
```diff
@@ -9,33 +9,37 @@ class PyBasemap(PythonPackage):
     depends_on("py-setuptools", type=("build", "run"))
     depends_on("py-numpy@1.2.1:", type=("build", "run"))
-    depends_on("py-matplotlib@1.0.0:3.0.0,3.0.2:", type=("build", "run"))
+    depends_on("py-matplotlib", type=("build", "run"))
```
Per `requirements.txt`

```
matplotlib >= 1.5, < 2.0; python_version == "2.6"
matplotlib >= 1.5, < 3.0; python_version == "2.7"
matplotlib >= 1.5, < 2.0; python_version == "3.2"
matplotlib >= 1.5, < 2.0; python_version == "3.3"
matplotlib >= 1.5, < 3.0; python_version == "3.4"
matplotlib >= 1.5, < 3.9; python_version >= "3.5"
```

and the python version range, this should be constrained:

```suggestion
    depends_on("py-matplotlib@1.5:3.8 ", type=("build", "run"))
```

LaurentChardon replied:
You're right, I'll change that, thanks.

tldahlgren commented on var/spack/repos/builtin/packages/py-basemap/package.py:
```diff
@@ -9,33 +9,37 @@ class PyBasemap(PythonPackage):
     # 1.2.1 is PROJ6 compatible
     # https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=939022
-    depends_on("py-pyproj@1.9.3:", type=("build", "run"), when="@1.2.1:")
+    depends_on("py-pyproj", type=("build", "run"))
```
This should also be constrained per `requirements.txt`:

```
pyproj >= 1.9.3, < 2.1.0; python_version == "2.6"
pyproj >= 1.9.3, < 2.2.0; python_version == "2.7"
pyproj >= 1.9.3, < 1.9.6; python_version == "3.2"
pyproj >= 1.9.3, < 2.1.0; python_version == "3.3"
pyproj >= 1.9.3, < 2.1.0; python_version == "3.4"
pyproj >= 1.9.3, < 3.7.0; python_version >= "3.5"
```

```suggestion
    depends_on("py-pyproj@1.9.3:3.6", type=("build", "run"))
```

LaurentChardon replied:
I'll change it

tldahlgren commented on var/spack/repos/builtin/packages/py-basemap/package.py:
```diff
@@ -9,33 +9,37 @@ class PyBasemap(PythonPackage):
-    depends_on("pil", type=("build", "run"))
-    depends_on("geos")
+    depends_on("py-cython", type=("build", "run"))
+    depends_on("py-pyshp", type=("build", "run"))
```
This should be constrained to match `requirements.txt`'s:

```
pyshp >= 1.2, < 2.0; python_version == "2.6"
pyshp >= 1.2, < 2.4; python_version >= "2.7"
```

LaurentChardon replied:
I'll change it

tldahlgren commented on var/spack/repos/builtin/packages/py-basemap/package.py:
```diff
@@ -9,33 +9,37 @@ class PyBasemap(PythonPackage):
-    depends_on("geos")
+    depends_on("py-cython", type=("build", "run"))
+    depends_on("py-pyshp", type=("build", "run"))
+    depends_on("geos", type=("build", "run"))
```
Looks like the dependency on `py-packaging` is missing, where `requirements.txt` has:

```
packaging >= 16.0, < 17.0; python_version == "2.6"
packaging >= 16.0, < 21.0; python_version == "2.7"
packaging >= 16.0, < 17.0; python_version == "3.2"
packaging >= 16.0, < 17.0; python_version == "3.3"
packaging >= 16.0, < 21.0; python_version == "3.4"
packaging >= 16.0, < 24.0; python_version >= "3.5"
```

LaurentChardon replied:
I'll add it

tldahlgren commented on var/spack/repos/builtin/packages/py-basemap/package.py:
```diff
@@ -9,33 +9,37 @@ class PyBasemap(PythonPackage):
-    depends_on("py-six", type=("build", "run"))
-    depends_on("pil", type=("build", "run"))
-    depends_on("geos")
+    depends_on("py-cython", type=("build", "run"))
```
Since the `py-cython` package includes some pretty old versions, this should be constrained according to `requirements-setup.txt`'s:

```
cython >= 0.29, < 3.1; python_version >= "3.3" or python_version < "3.0"
cython >= 0.26, < 0.27; python_version == "3.2"
```

LaurentChardon replied:
I'll change it, thanks

tldahlgren commented on var/spack/repos/builtin/packages/py-basemap/package.py:
```diff
@@ -9,33 +9,37 @@ class PyBasemap(PythonPackage):
     depends_on("c", type="build")  # generated
     depends_on("cxx", type="build")  # generated
 
+    depends_on("python@3.10:3.11", type=("build", "run"))
```
Curious. Where is the limit `3.11` coming from? 

`basemap_data/setup.py` indicates simply `<4` and the `python` package has up to `3.13.1` now. So would it be better to have a range of `@3.10:3` (to indicate any version from 3.10 on)?

tldahlgren replied:
Note: `baseman_data_hires/setup.py` indicates the same range.

LaurentChardon replied:
Yes it's better. I think I had it this way for testing purposes and forgot to remove it, but a tighter approach as you suggest is the way to go. I'll change it.

tldahlgren reviewed (commented):
Great start. I question removing the older version though it is pretty old. More importantly, it looks like there's a need for tighter version constraints.

I also recommend creating three packages and have `py-basemap` depend on `py-basemap-data` and, if `+hires`, `py-basemap-data-hires`.

tldahlgren reviewed (changes requested):
See previous comments.

LaurentChardon commented:
Thanks for your help @tldahlgren. I'll work on some of these changes.

LaurentChardon committed:
py-basemap: add v1.4.1

* py-basemap: remove broken v1.2.1
* py-basemap: add v1.4.1
* py-basemap: add hires variant

LaurentChardon commented:
I've pushed the following modifications:
- revert install method to python setup.py
- tighten dependencies versions
- moved dependecies in blocks for clarity
- Added missing packaging dependency

I did not split it in 3 packages, let me know if you're OK with that.
Thanks!

tldahlgren reviewed (approved):
Confirmed the version sha256. Attempted to build but am having problems with building one of the dependencies on my system.
